File: stack.go
package main

type SNode struct {
    value string
    next  *SNode
}

type Stack struct {
    name string
    top  *SNode
}

func createStack(name string) *Stack {
    return &Stack{
        name: name,
        top:  nil,
    }
}

func (s *Stack) push(value string) {
    node := &SNode{value: value, next: s.top}
    s.top = node
}

func (s *Stack) pop() string {
    if s.top == nil {
        return ""
    }
    value := s.top.value
    s.top = s.top.next
    return value
}

func (s *Stack) peek() string {
    if s.top == nil {
        return ""
    }
    return s.top.value
}

func (s *Stack) read() string {
    result := ""
    cur := s.top
    for cur != nil {
        result += cur.value
        if cur.next != nil {
            result += " "
        }
        cur = cur.next
    }
    return result
}

File: stack.cpp
#include "stack.h"
#include <string>
using namespace std;

Stack* create_stack(string name) {
    Stack* s = new Stack;
    s->name = name;
    s->top = nullptr;
    return s;
}

void stack_push(Stack* s, string value) {
    SNode* node = new SNode;
    node->value = value;
    node->next = s->top;
    s->top = node;
}

string stack_pop(Stack* s) {
    if (!s->top) return "";
    SNode* node = s->top;
    string value = node->value;
    s->top = node->next;
    delete node;
    return value;
}

string stack_peek(Stack* s) {
    if (!s->top) return "";
    return s->top->value;
}

string stack_read(Stack* s) {
    string out = "";
    SNode* cur = s->top;
    while (cur) {
        out += cur->value;
        if (cur->next) out += " ";
        cur = cur->next;
    }
    return out;
}

